doc: clarify test placement guidance (#35260)
bitcoin/bitcoin: doc/developer-notes.md

MacOS: `brew install doxygen graphviz`

> ## Tests and Commit Structure

Reviewer: I have doubts that allocating a full top-level ("##") section to this issue is the best. Maybe the section could be called "General Testing", with the first paragraph, and the second paragraph demoted by one level to "### Commit Structure for Tests".

Author: Done, thanks

Tests should be added or updated for new features, bug fixes, and other behavior
changes. A behavior-changing commit should include the relevant test coverage.
Test changes can be omitted for behavior-preserving work with no externally
observable behavior to test, such as moving code, renaming, or mechanical
refactors.

> 

Reviewer: The general points on testing could also include that tests should check results or post-conditions (a test without proper checks increases code coverage, but has limited value). But this is already outside of the scope of the original intent.

Author: Partially done

MacOS: `brew install doxygen graphviz`

## General Testing

As a rule of thumb, an externally observable behavior change (new feature, bug
fix, or changed default) should be accompanied by an automated test.
New tests are usually not needed for behavior-preserving work,
such as moving code, renaming, or mechanical refactors.
When an automated test would be brittle or have limited long-term value,
a manual testing recipe in the commit message or PR description is an
acceptable alternative.

### Commit Structure for Tests

> Tests typically go in the commit whose behavior they prove. For a single

Reviewer 2: nit: the previous sentence can be kept. 

```suggestion
Tests typically go in the commit whose behavior they prove. When 
existing tests already cover the behavior being changed, update 
those tests in the same commit as the behavior change. For a single
```

Author: Mostly done, let me know what you think.

behavior-changing commit, this placement can be verified by running the new
test against the old implementation and observing it fail. If the current
behavior is not already covered by tests, consider first adding a

> behavior-preserving test commit that documents it.

Reviewer 3: I’m not opposed to this paragraph, but IMO it would be sufficient to just fix the behavior and add a test for the correct behavior in the same change, rather than having one commit first asserting the incorrect behavior.

Author: > it would be sufficient to just fix the behavior and add a test for the correct behavior in the same change, rather than having one commit first asserting the incorrect behavior.

> a commit that fixes the problem, followed by a commit that enforces the correct behavior in a test

I agree this is often fine for simple fixes, and the text now says a simple feature or bug fix can add its test in the same commit.
The case I want to document is the non-trivial one, where the existing behavior is uncovered and the fix and test are intertwined. Adding the final test after the fix, or relying on reviewers to cherry-pick/revert code to make it fail, does not clearly record the smallest invariant being changed. It just documents the post-fix behavior without clarifying exactly what behavior was modified. It also makes review harder when reverting the fix is not clean or when the change is mixed with refactoring.

The split is meant to avoid that: first add a characterization test that records the current result, potentially marking wrong or temporary expectations with a `TODO`; then the behavior-changing commit removes the `TODO` and updates the assertion. That makes the diff show exactly which expectation changed, while still keeping each commit passing.

The reviewer or CI can just run both commits' tests to prove the exact effect of the behavior change, with no need for partial reverts unless that's simpler, which isn't always the case.

MacOS: `brew install doxygen graphviz`

## General Testing

> Tests should be added or updated for new features, bug fixes, and other behavior

Reviewer 4: Also, adding a test is not always the best choice. Sometimes it is better to manually document how to test something, instead of adding a brittle and confusing test for eternity.

Author: Covered in the latest push, added you as coauthor, thanks.

changes. A behavior-changing commit should include the relevant test coverage.
Test changes can be omitted for behavior-preserving work with no externally
observable behavior to test, such as moving code, renaming, or mechanical

> refactors.

Reviewer 5: I'm not sure about adding this. We already seem to be doing a decent job at it, and as maflcko pointed out, sometimes the tests being added for an obvious one-line change can be brittle, or just don't warrant the additional complexity. Some tests do also require non-trivial maintenance. Maybe this can be formulated in a bit more aspirational tone? I think you make a good job of that in the other paragraph, with "rule of thumb", "consider", etc.

Author: I have rephrased this 10 times again, let me know what you think.

### Commit Structure for Tests

A good rule of thumb is to put tests in the commit whose behavior they prove. In
a single behavior-changing commit, the relationship can often be checked by
running the new test against the old implementation and observing it fail. In a
stacked change, if the current behavior is not already covered and a later
commit intentionally changes it, consider first adding a behavior-preserving test
commit that documents the current behavior, even if that expectation will change.
This [characterization testing](https://en.wikipedia.org/wiki/Characterization_test)
style keeps the evidence more granular: the test-only commit records prior
behavior, and the behavior-changing commit shows the minimal expectation change;
see also [Michael Feathers' original
write-up](https://michaelfeathers.silvrback.com/characterization-testing).
When prior behavior is already covered, update those tests in the same commit as

> the behavior change.

Reviewer 4: Reminds me of:

```
CONTRIBUTING.md:125:This means tests must be updated in the same commit that changes the behavior.
```

Maybe that line can be replaced by a link to this section?

Not sure what should go in `CONTRIBUTING.md` vs this file here.

Author: Good point, replaced the `CONTRIBUTING.md` duplication with a link to the new developer-notes section.

Reviewer 6: Would expect these paragraphs in CONTRIBUTING.md, but I guess developer-notes.md already has advice on scripted diffs etc so this works too.

A few non-style-related recommendations for developers, as well as points to
pay attention to for reviewers of Bitcoin Core code.

> ## General Testing

Reviewer 5: Major Header 🫡

As a rule of thumb, an externally observable change (new feature, bug fix, or
changed default) should be accompanied by an automated test.
New tests are usually not needed for behavior-preserving work that is easy to
validate, such as moving code, renaming, or mechanical refactors.
When an automated test would be brittle or have limited long-term value,
a manual testing guide in the commit message or PR description is an
acceptable alternative.

### Commit Structure for Tests

Test placement depends on existing coverage and the type of change:

> 

Reviewer 7: ```suggestion

* All tests must pass at each individual commit; CI enforces this.
```
Maybe this is obvious (feel free to ignore this), but this wasn't clear to me initially. I discovered a bug, wrote a test to demonstrate the bug, made that the first commit, then fixed it in the second commit. It seemed logical to me, because it allows a reviewer to `git checkout` the first commit and verify that the bug exists. Then I was told that's not how we do things. (I looked at the entire doc just now and don't see this already mentioned.) I guess this suggestion is implied by your existing bullet points, but it might be good to mention it explicitly; that would have helped me.

Author: > allows a reviewer to git checkout the first commit and verify that the bug exists

See https://github.com/bitcoin/bitcoin/blob/31abaa264c35d1779a34d8c110a0c91e62c08ee0/CONTRIBUTING.md?plain=1#L107-L109

The characterization structure does even more: you don't have a failure and a fix, but you have a test that describes the current behavior with TODOs describing what the desired behavior should be, and a fix commit which only changes the lines that would fail otherwise - proven by the CI. So you don't even need to do the interactive rebase/checkout game, just run the test for each commit without needing to adjust the old code to work for the new test.

* When existing tests already cover the behavior being changed, update them in
  the same commit as the change. The diff records the old and new expectations
  together and shows the change was intentional.
* For a simple feature or bug fix without existing coverage, the change and its
  test can often be in the same commit.
* For a non-trivial refactor, if the relevant invariant is not already covered
  by automated tests, first add that coverage in a separate test commit. The
  refactor commit should not need to update test expectations.
* For a non-trivial change to existing behavior without coverage, consider
  adding a preceding commit with a [characterization test](https://en.wikipedia.org/wiki/Characterization_test)
  that covers the behavior before it changes. Add `TODO` comments next to

>   assertions that capture the incorrect results so they are not

Reviewer 8: nit (i think this part also reads better with these changes):
```suggestion
  to document the current behavior. Mark any assertions
  whose values will change with a `TODO` comment so they are not
```

Author: Done something similar, thanks.

  mistaken for intended behavior. The behavior-changing commit then updates

>   those assertions and removes the comments.

Reviewer 8: nit (last sentence sounds a bit mechanical):
```suggestion
  mistaken for intended behavior. In the behavior-changing commit remove the `TODO` comments when updating those assertions.
```

Author: Done something similar, thanks.
